Library/Homebrew/extend/fileutils.rb

```diff
@@ -6,42 +6,95 @@
 # @see http://ruby-doc.org/stdlib-1.8.7/libdoc/fileutils/rdoc/FileUtils.html Ruby's FileUtils API
 module FileUtils
   # Create a temporary directory then yield. When the block returns,
-  # recursively delete the temporary directory.
-  def mktemp(prefix = name)
-    prev = pwd
-    tmp  = Dir.mktmpdir(prefix, HOMEBREW_TEMP)
-
-    # Make sure files inside the temporary directory have the same group as the
-    # brew instance.
-    #
-    # Reference from `man 2 open`
-    # > When a new file is created, it is given the group of the directory which
-    # contains it.
-    group_id = if HOMEBREW_BREW_FILE.grpowned?
-      HOMEBREW_BREW_FILE.stat.gid
-    else
-      Process.gid
+  # recursively delete the temporary directory. Passing opts[:preserve]
+  # or calling mktemp_preserve during the block's execution will suppress
+  # the deletion for the topmost mktemp on the call stack, if any.
+  # Returns an Mktemp.
+  # Not thread-safe.
+  def mktemp(prefix = name, opts = {})
+    result = Mktemp.new(prefix, opts)
+    result.run do
+      yield
     end
-    begin
-      # group_id.to_s makes OS X 10.6.7 (ruby-1.8.7-p174) and earlier happy.
-      chown(nil, group_id.to_s, tmp)
-    rescue Errno::EPERM
-      opoo "Failed setting group \"#{Etc.getgrgid(group_id).name}\" on #{tmp}"
+    result
+  end
+  module_function :mktemp
+
+  # Tracks behavior of an mktemp() invocation
+  class MktempResult
+    attr_accessor :tmpdir, :preserve
+  end
+
+  # Performs mktemp's functionality, and tracks the results.
+  class Mktemp
+    include FileUtils
+
+    # Path to the tmpdir used in this run.
+    attr_accessor :tmpdir
+    # Whether the files in tmpdir were retained
+    attr_accessor :preserve
+    # Stack of currently-executing Mktemp objects
+    @@current_mktemp = []
+
+    # Activates tempdir preservation on the currently-executing run, if any.
+    # If called while multiple Mktemps are on the call stack, only the topmost
+    # (most recently invoked) one is affected.
+    def self.preserve_current_run
+      return if @@current_mktemp.empty?
+      @@current_mktemp.last.preserve = true
     end
 
-    begin
-      cd(tmp)
+    def initialize(prefix = name, opts = {})
+      @prefix = prefix
+      @preserve = opts[:preserve]
+    end
 
+    def run()
```

Comment:
It's customary to omit an empty argument list (even in a declaration like here).